tests/group_functions/group_scan.h

```diff
@@ -20,11 +20,149 @@
+          auto res_e_acc =
+              res_e_sycl.template get_access<sycl::access::mode::read_write>(
+                  cgh);
+          auto res_i_acc =
+              res_i_sycl.template get_access<sycl::access::mode::read_write>(
+                  cgh);
+          auto end_e_acc =
+              end_e_sycl.template get_access<sycl::access::mode::read_write>(
+                  cgh);
+          auto end_i_acc =
+              end_i_sycl.template get_access<sycl::access::mode::read_write>(
+                  cgh);
+          auto ret_type_e_acc =
+              ret_type_e_sycl
+                  .template get_access<sycl::access::mode::read_write>(cgh);
+          auto ret_type_i_acc =
+              ret_type_i_sycl
+                  .template get_access<sycl::access::mode::read_write>(cgh);
+
+          cgh.parallel_for<
+              joint_scan_group_kernel<D, T, U, Group, with_init, I, OpT>>(
+              executionRange, [=](sycl::nd_item<D> item) {
+                Group group = get_group<Group>(item);
+
+                T* v_begin = v_acc.get_pointer();
+                T* v_end = v_begin + v_acc.size();
+                U* r_e_begin = res_e_acc.get_pointer();
+                U* r_i_begin = res_i_acc.get_pointer();
+
+                auto r_e_end = joint_exclusive_scan_helper<with_init, I>(
+                    group, v_begin, v_end, r_e_begin, op);
+                ret_type_e_acc[0] = std::is_same_v<U*, decltype(r_e_end)>;
+
+                auto r_i_end = joint_inclusive_scan_helper<with_init, I>(
+                    group, v_begin, v_end, r_i_begin, op);
+                ret_type_i_acc[0] = std::is_same_v<U*, decltype(r_i_end)>;
+
+                end_e_acc[0] = (r_e_begin + res_e_acc.size() == r_e_end);
+                end_i_acc[0] = (r_i_begin + res_i_acc.size() == r_i_end);
+              });
+        })
+        .wait_and_throw();
+  }
+  CHECK(res_e_end);
+  CHECK(res_i_end);
+  CHECK(ret_type_e);
+  CHECK(ret_type_i);
+  std::vector<U> reference_e(size, U(-1));
+  std::vector<U> reference_i(size, U(-1));
+
+// FIXME: known_identity is not implemented in hipSYCL yet
+#if SYCL_CTS_COMPILING_WITH_HIPSYCL
+  I init_value = (with_init) ? I(init)
+                 : (std::is_same_v<OpT, sycl::plus<U>>)
+                     ? I{}
+                     : std::numeric_limits<I>::lowest();
+#else
+  I init_value = (with_init) ? I(init) : sycl::known_identity<OpT, I>::value;
+#endif
+  std::exclusive_scan(v.begin(), v.end(), reference_e.begin(), init_value, op);
+  std::inclusive_scan(v.begin(), v.end(), reference_i.begin(), op, init_value);
+  for (int i = 0; i < size; i++) {
+    {
+      INFO("Check joint_exclusive_scan for element " + std::to_string(i));
+      INFO("Result: " + std::to_string(res_e[i]));
+      INFO("Expected: " + std::to_string(reference_e[i]));
+      // ulp is i because there can be different rounding mods
+      CHECK(is_equal_with_ulp(res_e[i], reference_e[i], i));
```

Comment:
This check is not correct. I suppose you add `is_equal_with_ulp` to compare floating point values, but floating point addition is not associative and result is non-deterministic.
Consider following input: {10^9, -10^9, 10^{-9}}. The last element of output can be either 0 or 10^{-9}. I can't provide an input to create arbitrary ULP difference.

Probably we can add verification only for pre-defined input and my proposal with using random input doesn't fly.

What is the size of buffer here? Up to some limit the expected ULP must be 0 as there should be no rounding errors to represent numbers 1, 2, 3, ... Using `size` for allowed error in ULP seems too loose.

Reply:
In this test with size of following values: {5, work_group_size / 2, 3 * work_group_size}, where work_group_size is max size of work-group supported by device. For scan_over_group tests size is work_group_size.
Should we use some fixed number instead and remove comparing with ulp?

Reply:
AFAIK, some devices support work-group size > 1000. I don't see much difference between 10 (27 == 3x3x3) and 1000 elements if we to test scan functionality. I would use minimal work-group size required for functional testing and check that tested device supports that minimal size.

With regard to validation of floating point variant, I think we have to use pre-defined input to keep the verification code simple. According to my understanding, using `is_equal_with_ulp` for pre-defined input is unnecessary, but it can be used. We just need to be more accurate with reference values. In opinion, current values are too loose.

Reply:
Set max size for work-group as 12 using functionality of helper function util::work_group_range. With restricted size and pre-defined values 1, 2, 3, ... there are no rounding errors and comparing with ulp is unnecessary - it is removed.